### PR TITLE
plan, executor: improve hash join to support all join types

### PR DESCRIPTION
**Pull request by zz-jason**
1. improve `HashJoinExec` to support **Semi Join**, **Anti Semi Join**, **Left Outer Semi Join** and **Anti Left Outer Semi Join**
2. improve `joinResultGenerator`s
2. `HashSemiJoin` is retained to support `ApplyJoinExec`
3.  add `order by` to some tests to make unit test stable
4. split `ApplyJoinExec`, `HashSemiJoinExec`, `NestedLoopJoinExec` to separate files

### CODE REVIEW

**Comment by zz-jason** [comment]
@hanfei1991 @coocood PTAL

**Comment by coocood** [comment]
It's hard to review if you split/rename files and change the code in the same PR.
Can you just modify the code in this PR and split/rename files in the next PR?

**Comment by zz-jason** [comment]
@coocood updated, PTAL

File: executor/executor_test.go [hunk 1]
```diff
@@ -1332,10 +1332,10 @@ func (s *testSuite) TestGeneratedColumnRead(c *C) {
 	result.Check(testkit.Rows(`1 2 3 2`, `3 4 7 12`, `8 8 16 64`))
 
 	// Test generated column in subqueries.
-	result = tk.MustQuery(`SELECT * FROM test_gc_read t WHERE t.a not in (SELECT t.a FROM test_gc_read t where t.c > 5)`)
+	result = tk.MustQuery(`SELECT * FROM test_gc_read t WHERE t.a not in (SELECT t.a FROM test_gc_read t where t.c > 5) order by a`)
```
Comment (hanfei1991) [inline executor/executor_test.go]: We can right as `result.Sort().Check`

File: plan/physical_plans.go [hunk 2]
```diff
@@ -568,6 +566,16 @@ func buildSchema(p PhysicalPlan) {
 		} else {
 			p.SetSchema(expression.MergeSchema(p.Children()[x.outerIndex].Schema(), p.Children()[1-x.outerIndex].Schema()))
 		}
+	case *PhysicalHashJoin:
+		if x.JoinType == SemiJoin || x.JoinType == AntiSemiJoin {
```
Comment (coocood) [inline plan/physical_plans.go]: Can we extract a function like
```
joinSchema(tp JoinType, children []Plan, outerIndex int)
```

File: executor/executor_test.go (hunk 1 above)
Comment (zz-jason) [reply executor/executor_test.go]: this involves many result order changes, adding a `order by` in the tail of a query is the simplest way to make test happy.

Comment (hanfei1991) [reply executor/executor_test.go]: But this may change the plan of query...

File: executor/join_result_generators.go [hunk 3]
```diff
@@ -34,49 +34,68 @@ var (
 // joinResultGenerator is used to generate join results according the join type, see every implementor for detailed information.
 type joinResultGenerator interface {
 	// emitMatchedInners should be called when key in outer row is equal to key in every inner row.
-	emitMatchedInners(outer Row, inners []Row, resultBuffer []Row) ([]Row, error)
+	emitMatchedInners(outer Row, inners []Row, resultBuffer []Row) ([]Row, bool, error)
```
Comment (coocood) [inline executor/join_result_generators.go]: Need to update comment about the return bool.

Comment (zz-jason) [reply executor/join_result_generators.go]: done

File: executor/join.go [hunk 4]
```diff
@@ -30,81 +30,88 @@ import (
...
 
-// hashJoinCtx holds the variables needed to do a hash join in one of many concurrent goroutines.
-type hashJoinCtx struct {
-	bigFilter   expression.CNFExprs
-	otherFilter expression.CNFExprs
-	// datumBuffer is used for encode hash keys.
-	datumBuffer   []types.Datum
-	hashKeyBuffer []byte
+type hashJoinBuffer struct {
+	datums []types.Datum
```
Comment (hanfei1991) [inline executor/join.go]: datums -> data

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**